barretenberg/cpp/src/barretenberg/vm/avm_trace/AvmMini_execution.cpp

```diff
@@ -0,0 +1,227 @@
+#include "AvmMini_execution.hpp"
+#include "barretenberg/common/serialize.hpp"
+#include "barretenberg/proof_system/circuit_builder/generated/AvmMini_circuit_builder.hpp"
+#include "barretenberg/vm/avm_trace/AvmMini_common.hpp"
+#include "barretenberg/vm/avm_trace/AvmMini_instructions.hpp"
+#include "barretenberg/vm/avm_trace/AvmMini_opcode.hpp"
+#include "barretenberg/vm/avm_trace/AvmMini_trace.hpp"
+#include "barretenberg/vm/generated/AvmMini_composer.hpp"
+#include <cstddef>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace avm_trace {
+
+/**
+ * @brief Run the bytecode, generate the corresponding execution trace and prove the correctness
+ *        of the execution of the supplied bytecode.
+ *
+ * @param bytecode A vector of bytes representing the bytecode to execute.
+ * @param calldata expressed as a vector of finite field elements.
+ * @throws runtime_error exception when the bytecode is invalid.
+ * @return A zk proof of the execution.
+ */
+plonk::proof Execution::run_and_prove(std::vector<uint8_t> const& bytecode, std::vector<FF> const& calldata)
+{
+    auto instructions = parse(bytecode);
+    auto trace = gen_trace(instructions, calldata);
+    auto circuit_builder = bb::AvmMiniCircuitBuilder();
+    circuit_builder.set_trace(std::move(trace));
+
+    auto composer = bb::honk::AvmMiniComposer();
+    auto prover = composer.create_prover(circuit_builder);
+    return prover.construct_proof();
+}
+
+/**
+ * @brief Parsing of the supplied bytecode into a vector of instructions. It essentially
+ *        checks that each opcode value is in the defined range and extracts the operands
+ *        for each opcode.
+ *
+ * @param bytecode The bytecode to be parsed as a vector of bytes/uint8_t
+ * @throws runtime_error exception when the bytecode is invalid.
+ * @return Vector of instructions
+ */
+std::vector<Instruction> Execution::parse(std::vector<uint8_t> const& bytecode)
+{
+    std::vector<Instruction> instructions;
+    size_t pos = 0;
+    const auto length = bytecode.size();
+
+    static_assert(sizeof(uint32_t) / sizeof(uint8_t) == AVM_OPERAND_BYTE_LENGTH);
+
+    while (pos < length) {
+        const uint8_t opcode_byte = bytecode.at(pos);
+        pos += AVM_OPCODE_BYTE_LENGTH;
+
+        if (!Bytecode::is_valid(opcode_byte)) {
+            throw std::runtime_error("Invalid opcode byte: " + std::to_string(opcode_byte));
+        }
+
+        const auto opcode = static_cast<OpCode>(opcode_byte);
+        auto in_tag_u8 = static_cast<uint8_t>(AvmMemoryTag::U0);
```

Comment:
Whats the casting for, is u8 = 0 not enough? 
question asked as a cpp noob

Reply:
The goal here is to keep in sync with the value defined in the enum. Sure, normally the enum value for U0 should always be 0, but if we decide to change the above code will become invalid.

Even though the enum is defined over uint8_t, the static cast is required otherwise compiler screams.